### PR TITLE
Puppet fact additions puppet v4

### PR DESCRIPTION
**Pull request by alexo1088**
Hi there.  Thanks for all the work performed on this repo!  

We have a client that's been utilizing this repo for quite some time.  When we originally implemented this, we noticed that data that was being collected kept using the Puppet master for the 'AWSUniqueId' instead of the instance tag itself.  Seems to be an issue with how the get_dimensions.rb file was performing the metadata lookup.  

Nonetheless, I re-wrote a small portion on this branch that has fixed the problem for our client (originally done on a fork).  I've since updated it to support v4 through your latest releases, and thought it would be a good idea to provide you with this info in case you'd like to merge into a new branch.  

Let me know what you think.

Thanks!

### CODE REVIEW

File: lib/facter/metadata_retrieval.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+begin
+  http.start
+  response = http.request(Net::HTTP::Get.new(uri.request_uri))
+rescue StandardError
+  raise 'Unable to reach AWS metadata'
+end
+unless response.nil? || response == 0
+  Facter.add('ec2_instance_identity_document') do
+    setcode do
+      JSON.parse(response.body).to_json
```
Comment (keitwb) [inline lib/facter/metadata_retrieval.rb:16]: Why does this get parsed from JSON and then put back into JSON, and then reparsed as JSON down in the function that uses it below?

File: lib/facter/metadata_retrieval.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+require 'net/https'
+uri = URI.parse('http://169.254.169.254/2014-11-05/dynamic/instance-identity/document')
+http = Net::HTTP.new(uri.host, uri.port)
+http.open_timeout = 4
+http.read_timeout = 4
+begin
+  http.start
+  response = http.request(Net::HTTP::Get.new(uri.request_uri))
+rescue StandardError
+  raise 'Unable to reach AWS metadata'
```
Comment (keitwb) [inline lib/facter/metadata_retrieval.rb:11]: Perhaps print the exception that was caught (unless Ruby already automagically does this) to give a little more detail to why the metadata fetch failed.

**Review by keitwb — COMMENTED** [review]
Hi, thanks for the code!  Just a few comments from me, but looks good overall.

Be on the lookout for our new agent which is coming out very soon which will supersede all of this (it actually does the AWS metadata discovery internally so external installers won't have to).

**Comment by atoulme** [comment]
Closing for inactivity. Please reopen if these changes are still needed.

**Pull request closed without merging**